### PR TITLE
git-branchless: fix build

### LINKED ISSUE #338971: Build failure: git-branchless
### Steps To Reproduce

1. build *git-branchless*

### Build log

https://hydra.nixos.org/search?query=git-branchless

### Additional context

No longer builds with latest Nixpkgs

### Notify maintainers

@arxanas

### Metadata

Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
$ nix run nixpkgs#nix-info -- "-m"
 - system: `"x86_64-linux"`
 - host os: `Linux 6.10.6, NixOS, 24.11 (Vicuna), 24.11.20240828.71e91c4`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Lix, like Nix) 2.92.0-dev-pre20240813-f9a3bf6
System type: x86_64-linux
Additional system types: i686-linux, x86_64-v1-linux, x86_64-v2-linux, x86_64-v3-linux
Features: gc, signed-caches
System configuration file: /etc/nix/nix.conf
User configuration files: /home/toastal/.config/nix/nix.conf:/etc/xdg/nix/nix.conf:/home/toastal/.nix-profile/etc/xdg/nix/nix.conf:/nix/profile/etc/xdg/nix/nix.conf:/home/toastal/.local/state/nix/profile/etc/xdg/nix/nix.conf:/etc/profiles/per-user/toastal/etc/xdg/nix/nix.conf:/nix/var/nix/profiles/default/etc/xdg/nix/nix.conf:/run/current-system/sw/etc/xdg/nix/nix.conf
Store directory: /nix/store
State directory: /nix/var/nix
Data directory: /nix/store/3g35mlki1l0p3xx68q6gpwpi5xavgrky-lix-2.92.0-dev-pre20240813-f9a3bf6/share`
 - nixpkgs: `/nix/store/887hpp8a2i99n9jjwcvz6qkhhhqsvzkg-source`
```

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

[^1]

[^1]: Please consider [giving up MS GitHub](https://sfconservancy.org/GiveUpGitHub/) or offering a non-proprietary, non-US-corporate-controlled mirror for this free software project. I wish to delete this Microsoft account in the future, but I need more projects like this to support alternative methods to send patches & contribute.

### PR DESCRIPTION
**Pull request by luftmensch-luftmensch**
## Description of changes
Closes #338971

Added new tests to skip that fails in an isolated environment

## Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Review by eclairevoyant — DISMISSED** [review]
Please keep any reformatting as its own commit

**Comment by luftmensch-luftmensch** [comment]
@eclairevoyant Done

File: pkgs/applications/version-management/git-branchless/default.nix [hunk 1]
```diff
@@ -50,13 +53,34 @@ rustPlatform.buildRustPackage rec {
     "--skip=test_switch_pty"
     "--skip=test_next_ambiguous_interactive"
     "--skip=test_switch_auto_switch_interactive"
+    "--skip=test_amend_undo"
+    "--skip=test_switch_pty"
+    "--skip=test_next_ambiguous_interactive"
+    "--skip=test_switch_auto_switch_interactive"
+    "--skip=test_amend_undo"
```
Comment (hmenke) [inline pkgs/applications/version-management/git-branchless/default.nix:60]: Duplicate
```suggestion
```

**Review by hmenke — CHANGES_REQUESTED** [review]
Why are there tests that are not `_pty` or `_interactive` in there? These shouldn't fail.

**Comment by niklaskorz** [comment]
The test output seems to be locale-specific, thus failing on non-English systems. Is there some environment variable we can pass so the tests are always run in English?

```
────────────┬───────────────────────────────────────────────────────────────────
    0       │-HEAD detached from 96d1c37
    1       │-Changes to be committed:
    2       │-  (use "git restore --staged <file>..." to unstage)
    3       │-  new file:   test3.txt
          0 │+HEAD losgelöst von 96d1c37
          1 │+Zum Commit vorgemerkte Änderungen:
          2 │+  (benutzen Sie "git restore --staged <Datei>..." zum Entfernen aus der Staging-Area)
          3 │+  neue Datei:     test3.txt
    4     4 │
    5       │-Changes to be committed:
          5 │+Zum Commit vorgemerkte Änderungen:
    6     6 │ diff --git c/test3.txt i/test3.txt
    7     7 │ new file mode 100644
    8     8 │ index 0000000..4706e16
    9     9 │ --- /dev/null
────────────┴───────────────────────────────────────────────────────────────────
To update snapshots run `cargo insta review`
Stopped on the first failure. Run `cargo insta test` to run all snapshots.
The application panicked (crashed).
Message:  snapshot assertion for 'amend_with_children-3' failed in line 45
Location: /private/tmp/nix-build-git-branchless-0.9.0.drv-0/git-branchless-0.9.0-vendor.tar.gz/insta/src/runtime.rs:563

Backtrace omitted. Run with RUST_BACKTRACE=1 environment variable to display it.
Run with RUST_BACKTRACE=full to include source snippets.


failures:
    test_amend_merge
    test_amend_with_children
    test_amend_with_working_copy

test result: FAILED. 12 passed; 3 failed; 0 ignored; 0 measured; 2 filtered out; finished in 6.94s

error: test failed, to rerun pass `--test test_amend`
```

**Comment by luftmensch-luftmensch** [comment]
@niklaskorz I don't know if this is a valuable option but we could do something like:

```nix
preCheck = lib.optionals stdenv.isDarwin ''
  # Need an UTF-8 locale for test_I test.
  export LANG=en_US.UTF-8
'';
```

Any thoughts on this approach?

**Comment by alixinne** [comment]
> Why are there tests that are not `_pty` or `_interactive` in there? These shouldn't fail.

I did run a `git bisect` the other day and it started failing with the update of Git in nixos-unstable to 2.46.0. Users seem to be reporting issues with Git 2.46.0 as well, so I'm not sure we should disable those tests indeed:
- https://github.com/arxanas/git-branchless/issues/1388
- https://github.com/arxanas/git-branchless/issues/1381

Those issues don't seem to have been traced back to Git 2.46.0 yet, but "updates to Git" are mentioned in https://github.com/arxanas/git-branchless/issues/1321#issuecomment-2125636129

**Comment by niklaskorz** [comment]
> @niklaskorz I don't know if this is a valuable option but we could do something like:
> 
> ```nix
> preCheck = lib.optionals stdenv.isDarwin ''
>   # Need an UTF-8 locale for test_I test.
>   export LANG=en_US.UTF-8
> '';
> ```
> 
> Any thoughts on this approach?

That doesn't seem to change anything, but don't worry about it. If I find out _why_ darwin git would be locale-dependent here and still ignore the LANG env var, I will open another PR.

**Review by samueltardieu — APPROVED** [review]

**Comment by samueltardieu** [comment]
Result of `nixpkgs-review pr 339023` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>git-branchless</li>
  </ul>
</details>

**Comment by hmenke** [comment]
@Artturin Why was this merged, even though I (one of the maintainers) requested changes that have *not* been addressed?

**Comment by Artturin** [comment]
> @Artturin Why was this merged, even though I (one of the maintainers) requested changes that have _not_ been addressed?

Sorry, my bad. I suppose I processed the top level review comment differently from a line review and did not think enough.

Here's a revert of the `fix build` commit with the formatting commit left alone https://github.com/NixOS/nixpkgs/pull/340854

**Comment by eclairevoyant** [comment]
Reverted - this probably merits discussion under the issue #338971 itself.

**Comment by bryango** [comment]
I am trying to revive this in #342278, please check out the comments there! Thanks!

**Pull request merged**